### PR TITLE
Eliminate cache requirement (resolves #586)

### LINKED ISSUE #586: Eliminate cache requirement
Resulting from a conversation with @arkal and @jvivian.

@arkal can explain in more detail but he's running into issues where cached files are evicted prematurely. All input, output and intermediate files involved in a single sample run of his pipeline can fit onto the disk space of a single beefy EC2 instance, yet the current caching implementation doesn't utilize that.

The cache should be shared by all concurrent workers on a machine. It should fill up the disk and evict files as late as possible. A simple heuristic for this would be to persistently track the disk requirements of the set of all concurrently running jobs on a machine and ensuring that a job's required disk space is available before it is added to that set, i.e. shortly before it is started. If not enough free space is available, files need to be evicted from the cache until it is. The persistent tracking could be done in a file with locking. The overhead of managing that file and the contention introduced by the file locking should both be negligible for the job throughput Toil currently achieves.

### PR DESCRIPTION
**Pull request by arkal**
Resolves #586

### CODE REVIEW

File: src/toil/common.py [hunk 1]
```diff
@@ -150,7 +150,14 @@ def fC(minValue, maxValue=None):
         setOption("defaultMemory", h2b, iC(1))
         setOption("defaultCores", float, fC(1.0))
         setOption("defaultDisk", h2b, iC(1))
-        setOption("defaultCache", h2b, iC(0))
+        #  defaultCache needs to be handled a tad differently because it can be either a float b/w [0.0, 1.0] or an int.
+        x = getattr(options, 'defaultCache', None)
```
Comment (hannes-ucsc) [inline src/toil/common.py]: options.defaultCache

**Comment by arkal** [comment]
The error was trivial. Need to invoke Job.CachedFileStore.terminateEvent.set() in case of failed cached run instead of Job.FileStore.terminateEvent.set(). It's fixed in the new commit. TOIL no longer hangs on raised error.

**Comment by hannes-ucsc** [comment]
I filed a PR against your fork with the changes that we did together.

Other things that should be addressed, in order of importance:

- Rebase now (and keep rebasing) to latest master
- Fix the Jenkins build
- The tests should be methods of a subclass of toil.test.ToilTest.

**Comment by hannes-ucsc** [comment]
The tests leave the working copy dirty. See end of build output. And keep this branch rebased onto master.

**Comment by hannes-ucsc** [comment]
bd2k-python-lib 1.12.dev8 now has an alternative to shutil.copyfileobj that support a limit:

https://github.com/BD2KGenomics/bd2k-python-lib/commit/7bba82990deb7dd65776bfb172c788e9310ee0c8

You'll want to update setup.py to refer to that version of bd2k-python-lib.

**Comment by arkal** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
The history of this branch is still a mess. Master has progress a bit and there a lot of "back-and-forth" commits in this branch. It also seems that this branch was never rebased? I did a trial rebase and hit a lot of conflicts. At this stage I recommend that you start a fresh branch off of master. Before you do that, make sure your master reflects the very latest upstream master (I think in your local clone that would be "bd2k/master"). Then check out your fresh branch, compare it (with PyCharm) with your issues/586a-eliminate-cache-requirement branch and apply the changes interactively in the split-pane diff editor.

From then on, be very eager to keep your fresh branch rebased on top of master. Do not merge from master. Use rebase instead.

**Comment by arkal** [comment]
Jenkins, test this please.

**Comment by arkal** [comment]
moved to #656 with a cleaner history.

**Pull request closed without merging**